src/sqlfluff/rules/L003.py

```diff
@@ -776,136 +830,263 @@ def _process_current_line(
-        cls, elem: BaseSegment, templated_file: Optional[TemplatedFile]
-    ) -> Optional[str]:
-        if elem.is_type("placeholder"):
-            if templated_file is None:
-                raise ValueError("Parameter templated_file cannot be: None.")
-            assert elem.pos_marker
-            slices = templated_file.raw_slices_spanning_source_slice(
-                elem.pos_marker.source_slice
+            memory["problem_lines"].append(n)
+
+        if fixes_list:
+            return LintResult(
+                anchor=anchor,
+                memory=memory,
+                description="Comment not aligned with following line.",
+                fixes=fixes_list,
             )
-            if slices:
-                return slices[0].slice_type
-        return None
 
-    @classmethod
-    def _single_placeholder_line(cls, current_line):
+        # NB: At shallower indents, we don't check, we just check the
+        # previous lines with the same balance. Deeper indents can check
+        # themselves.
+        return LintResult(memory=memory)
+
+
+class _TemplateLineInterpreter:
+    def __init__(
+        self,
+        current_line: List[BaseSegment],
+        templated_file: Optional[TemplatedFile],
+    ) -> None:
+        self.current_line = [el for el in current_line if not el.is_whitespace]
+        self.templated_file = templated_file
+
+    @property
+    def template_content(self):
+        return "".join(
+            seg.raw or getattr(seg, "source_str", "") for seg in self.current_line
+        )
+
+    def is_single_placeholder_line(self):
         count_placeholder = 0
-        for seg in current_line:
+        for seg in self.current_line:
             if seg.is_code:
                 return False
             elif seg.is_type("placeholder"):
                 count_placeholder += 1
         return count_placeholder == 1
 
-    @classmethod
-    def _is_template_block_start_line(cls, current_line, templated_file):
-        def segment_info(idx: int) -> Tuple[str, Optional[str]]:
-            """Helper function."""
-            seg = current_line[idx]
-            return seg.type, cls._get_element_template_info(seg, templated_file)
-
-        if not cls._single_placeholder_line(current_line):
-            return False
-        for idx in range(1, len(current_line)):
-            if segment_info(idx - 1) in (
-                ("placeholder", "block_start"),
-                ("placeholder", "compound"),
-                ("placeholder", "block_mid"),
-            ) and segment_info(idx) == ("indent", None):
-                return True
-        return False
+    def list_segement_and_raw_segement_types(self):
+        """Yields the tuple of seg type and underlying type were applicable."""
+        for seg in self.current_line:
+            raw_seg = self.get_raw_slices(seg)
+            yield (seg.type, raw_seg[0].slice_type if raw_seg else None)
+
+    def iterate_type_pairs(self):
```

Comment:
Maybe rename for clarity?
```suggestion
    def iterate_adjacent_type_pairs(self):
```